## Chou-Fasman

Python implementation of Chou-Fasman algorithm for protein secondary structure prediction.

Chou-Fasman algorithm was the first protein structure prediction algorithm developed by Chou and Fasman in 1974. It pioneered a series of developments in protein structure prediction. This algorithm is based on a propensity table that was constructed by examining the x-ray-determined structures of 15 proteins containing 2473 amino acid residues and the number of occurrences of a given amino acid in the alpha-helix, beta-sheet, and coil was tabulated. The algorithm uses the propensity table and the set of empirical rules formulated by Chou and Fasman for protein secondary structure prediction.

## GOR

GOR (Garnier–Osguthorpe–Robson) is a protein secondary structure prediction method developed in late 1970s shortly after Chou–Fasman method. This method is also based on the probability tables constructed examining the information on protein structures derieved from X-ray crystallography but it also takes into consideration the probability values of immediate neighbors. GOR method predict alpha helix, beta sheet, turn, or random coil at each position based on 17-amino-acid sequence windows. 

## PhyloBit

PhyloBit is a phylogenetic tree construction tool that is based on Distance-Matrix method for making tree. It use JC69 nucleotide substitution model (Jukes and Cantor 1969) and takes aligned multi fasta file as an argument. PhyloBit is available for both windows and linux distribution in command-line and GUI interface.

## GC_Classifier

Check the genomic GC profile based similarity between the two fasta record (Reference genome and Test genome) and plot the data to show the relationship between the two genome.
 
## Seq-Logo

Python implementation for creating nucleotide and amino acid sequence logo from multi fasta file. 

## GFF_Extractor

Working on ...

## Nussinov

Working on ...
